tests/marine/benchmark/conversation_generators.py

```diff
@@ -0,0 +1,165 @@
+from functools import wraps
+from random import randint
+from secrets import token_bytes
+from typing import List, Callable
+
+from pypacker.layer12 import ethernet
+from pypacker.layer3 import ip
+from pypacker.layer4 import tcp, udp
+from pypacker.pypacker import Packet
+
+from tests.marine.benchmark.utils import IpPair, BenchmarkPacket
+
+
+def udp_base_layer(
+    fn: Callable[[Packet, Packet, int], List[BenchmarkPacket]]
+) -> Callable[[IpPair, int], List[BenchmarkPacket]]:
+    @wraps(fn)
```

Comment:
This entire file is way more comlicated than it has any right to.
Why is everything a decorator and not a simple function? Your decorators don't really treat your functions as black boxes.

Reply:
What isn't treated as a blackbox? The decorated function simply returns an instance of `BenchmarkPacket`.
The "complication" stems from the generation of conversations - I wanted it to be easy to generate layer-5 conversations (notice how easy it will be to generate an HTTP conversation now - it will simply call existing decorators and return `BenchmarkPacket` with HTTP init)

Reply:
You can do all that with plain functions, without decorators that completely change the function signatures.
I think this structure will be way clearer:
```python
def _amplify_to_conversartion(side_a_packet: Packet, side_b_packet: Packet, conversation_length: int) -> Iterable[Packet]: ...

def _generate_sides_tcp_packets(ip_pair: IpPair) -> Tuple[Packet, Packet]: ...
def _generate_sides_udp_packets(ip_pair: IpPair) -> Tuple[Packet, Packet]: ...

def generate_raw_tcp_conversation(ip_pair: IpPair, conversation_length: int) -> Iterable[BenchmarkPacket]:
    side_a_packet, side_b_packet = _generate_sides_tcp_packets(ip_pair)
    packets = _amplify_to_conversation(side_a_packet, side_b_packet, conversation_length)
    for packet in packets:
        yield BenchmarkPacket(...) 
```